From 71dbf87aac84518249e7bd5973a32ae1ae055f74 Mon Sep 17 00:00:00 2001
From: Rob Evers <revers@redhat.com>
Date: Wed, 25 Jun 2014 14:11:40 -0400
Subject: [scsi] be2iscsi: Fix destroy MCC-CQ before MCC-EQ is destroyed

Message-id: <1403705504-27660-7-git-send-email-revers@redhat.com>
Patchwork-id: 84800
O-Subject: [RHEL7.1 e-stor PATCH 06/10] scsi: be2iscsi: Fix destroy MCC-CQ before MCC-EQ is destroyed
Bugzilla: 1088576
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>

From: John Soni Jose <sony.john-n@emulex.com>

http://www.spinics.net/lists/linux-scsi/msg74147.html

 EQ teardown should happen only after all CQ are destroyed.
 In some FW config, adapter goes into a freeze state. This
 fix moves teardown of MCC-Q before the EQ teardown happens.

Signed-off-by: John Soni Jose <sony.john-n@emulex.com>
Signed-off-by: Jayamohan Kallickal <jayamohan.kallickal@emulex.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/be2iscsi/be_main.c b/drivers/scsi/be2iscsi/be_main.c
index eac7b43..57605ef 100644
--- a/drivers/scsi/be2iscsi/be_main.c
+++ b/drivers/scsi/be2iscsi/be_main.c
@@ -3735,7 +3735,7 @@ static void hwi_cleanup(struct beiscsi_hba *phba)
  struct hwi_controller *phwi_ctrlr;
  struct hwi_context_memory *phwi_context;
  struct hwi_async_pdu_context *pasync_ctx;
- int i, eq_num, ulp_num;
+ int i, eq_for_mcc, ulp_num;
 
  phwi_ctrlr = phba->phwi_ctrlr;
  phwi_context = phwi_ctrlr->phwi_ctxt;
@@ -3772,16 +3772,17 @@ static void hwi_cleanup(struct beiscsi_hba *phba)
   if (q->created)
    beiscsi_cmd_q_destroy(ctrl, q, QTYPE_CQ);
  }
+
+ be_mcc_queues_destroy(phba);
  if (phba->msix_enabled)
-  eq_num = 1;
+  eq_for_mcc = 1;
  else
-  eq_num = 0;
- for (i = 0; i < (phba->num_cpus + eq_num); i++) {
+  eq_for_mcc = 0;
+ for (i = 0; i < (phba->num_cpus + eq_for_mcc); i++) {
   q = &phwi_context->be_eq[i].q;
   if (q->created)
    beiscsi_cmd_q_destroy(ctrl, q, QTYPE_EQ);
  }
- be_mcc_queues_destroy(phba);
  be_cmd_fw_uninit(ctrl);
 }
 
-- 
1.7.1